test/EntityFramework/FunctionalTests/PlanCompiler/LimitExpressionTests.cs

```diff
@@ -834,6 +836,42 @@ orderby building.Address.ZipCode
         }
 
         [Fact]
+        public void Limit_ComplexModel_OrderBy_Skip_without_rownumber()
+        {
+            var log = new StringWriter();
+            using (var context = new AdvancedPatternsMasterContext())
+            {
+                var previousFlag = SqlProviderServices.UseOrderByRowNumberForQueriesWithOffset;
+                try
+                {
+                    SqlProviderServices.UseOrderByRowNumberForQueriesWithOffset = false;
```

Comment:
Re query caching we could consider setting a separate "locking" internal flag the first time the value of the flag is used. That way you should still be able to set it very early before you start issuing queries but it would throw if you change it later. 

That said, I am not too concerned because this is very advanced.

cc @ajcvickers